### ISSUE TITLE
Parallel computations in complex testcases crash randomly in the function isoCut*::quadAreaCoeffs

### ISSUE THREAD

**Issue by GonzaloSaez** [isoAdvector/isoAdvector#7]
Dear Mr. Roenby,

Sorry to bother you again with another issue, but this one might be of your interested. I'm interested in applying the isoAdvector solver in complex drop impact on walls and liquid films testcases. I have been able of comparing isoAdvector and MULES in a drop impact on a big liquid film and I have observed that isoAdvector is capable of capturing the vortex rings and Ejecta which appear during the first instants of the impact (also captured with GERRIS solver). MULES was not able of capturing any of those phenomena.

These simulations require paralellisation, as cell count is high. However, when running these simulations in parallel, sigFPE errors concerning the _isoCut*::quadAreaCoeffs_ (where * stands for Cell and Face) function are found. The time at which these errors appear is not repeatable, meaning that they can arise or not during the computation. Normally a restart from a previous timestep allows to skip the errors, but sooner or later they appear again. An important issue of these problems is that they seem to have a random nature and that they are not fixed to given timesteps (non-repeatability). Moreover, when the errors are overcome the simulation results don't seem to give nonphysical values meaning that it shouldn't be a divergence error coming from the solver.

As they are sigFPE erros I suppose they come from divisions by zero. The only part of the code where we I find the possibility of having these zero divisions in the _quadAreaCoeffs_ function is the following:

```
        vector xhat = B - A;
        xhat /= mag(xhat);
        vector yhat = D - A;
        yhat -= ((yhat & xhat)*xhat);
        yhat /= mag(yhat);
```
Maybe a adding _SMALL_ to the computations of the magnitude of the vectors could help to reduce the issue, but the fact that the errors are random might imply that the issue is deeper into the code with paralellisation or numerical noise playing an important role.

In case you are interested in trying to debug these issues I could upload the testcases where I find these issues. As I suppose you wouldn't want to debug this issue in cases with 0.5M cells I shall reduce the cell count and see if the errors can also be found before other numerical issues appear due to coarsening.

Thank you for your help

(Our team is also interested in trying to expand and implement your formulation to 3 or more phase-flows, is it possible to discuss with you this topic?)

**Comment by isoAdvector** [comment]
1. You are not bothering me at all. I am grateful for any input that can help in the testing and maturing of isoAdvector
2. Extremely motivating to hear that you can actually capture some physics with isoAdvector that you couldn't with MULES. If you have any animations or pictures I'd love to see them.
3. Please send me a case that crashes including everything that is needed to run it (Allrun script etc), and I will have a look at it.
4. Could you do me a favor and compile the isoAdvector develop branch, add "gradAlphaNormal true;" to the isoAdvector subDict in fvSolution and see if this prevents crashing (See README for explanation). 
5. Also make sure that your p_rghFinal tolerance is well below the surfCellTol. Else you may be isoCutting non-surface cells which quadAreaCoeffs does not like.
6. Geometric VOF with 3 phases sounds interesting but difficult. How big is your research budget? :-) Let us discuss it via e-mail: jro@dhigroup.com

**Comment by GonzaloSaez** [comment]
I have just sent you an email with the testcase and the comparison you asked for. Hopefully we can solve these issues!

**Comment by isoAdvector** [comment]
Hello

For the moment, I am busy preparing isoAdvector for release in OpenFOAM-v1706+ over at 

https://develop.openfoam.com/Community/Integration-isoAdvector

In that version of isoAdvector, I have modified isoCutFace::cutPoints and isoCutFace::quadAreaCoeffs to "to avoid/treat consistently situations with a triangular rather than a quadrilateral area swept by FIIL. New implementation tested with damBreak, discInUniFlow on hex and sphereInUniFlow. Seems to work as it should." (Commit d95e2598). I have reason to believe that this modification will solve your problem. Would you mind trying this out and let me know if it works? For the purpose of testing, I think you should be able to simply copy the body of the two mentioned functions and overwrite what you have (However, there might also be other relevant fixes in the new version).

Best regards,
Johan

**Comment by GonzaloSaez** [comment]
Hi Roenby,

I saw today the commit and I answered in the issue posted in the develop.openfoam.com site. I copied the functions you mentioned in isoCutFace.C and it seems that for the smaller cases the random issue is gone. I will try these modifications in the parallel complex testcases and I will let you know if the random errors persist. 

By the way due to issues with the email servers in the past month, a lot of emails didn't reach my mailbox so if I didn't answer to any of your emails it was due to this issue, sorry for that.

Best
